### PR TITLE
feat(code-mappings):  Add support for stacktrace frames with backslashes for automatic code mappings

### PR DESCRIPTION
**Pull request by MichaelSun48**
This PR updates the existing code mapping logic to support stack traces whose filepaths contained backslashes instead of forward slashes. This effectively adds automatic code mapping support for issues derived from Windows machines (for the existing supported languages).

I've made the choice to save stack roots in the code mappings using whichever slash delimiter that the original stack path used. For example, if we have the following stack path and source:

* Stack Path = `"D:\Users\code\sentry\models\release.py"`
* Source path = `"src/sentry/models/release.py"`

Then the roots generated will be:

* Stack Root = `"D:\Users\code\"` (note that this source root still uses backslashes instead of forward slashes)
* Source Root = `"src/"`

This should be less confusing to the end user since we explicitly say we are going to replace the `stack_root` in the `stack_path` with the `source_root` to get the `source_path`. It might be confusing if the `stack_root`'s slash-delimiter was not the same as that in the `stack_path`. 

This also means that the previous logic to apply code mappings to a given frame to generate a stacktrace link needed to be updated so that the source path generated would only include slashes (since this path is used in a github URL, which only uses forward slashes).

### CODE REVIEW

File: src/sentry/integrations/utils/code_mapping.py [hunk 1]
```diff
@@ -61,15 +61,19 @@ class UnsupportedFrameFilename(Exception):
 class FrameFilename:
     def __init__(self, frame_file_path: str) -> None:
         self.raw_path = frame_file_path
-        if frame_file_path[0] == "/":
+        is_windows_path = False
+        if "\\" in frame_file_path:
```
Comment (bruno-garcia) [inline src/sentry/integrations/utils/code_mapping.py:65]: can this conflict with the escape character on Linux? e.g: `\` as an escape for a space in the path? or is that just a shell thing?

File: src/sentry/integrations/utils/code_mapping.py [hunk 2]
```diff
@@ -370,6 +386,10 @@ def convert_stacktrace_frame_path_to_source_path(
     If the code mapping does not apply to the frame, returns None.
     """
 
+    stack_root = code_mapping.stack_root
+    if "\\" in code_mapping.stack_root:
+        stack_root = code_mapping.stack_root.replace("\\", "/")
+
```
Comment (MichaelSun48) [inline src/sentry/integrations/utils/code_mapping.py:392]: This function (`convert_stacktrace_frame_path_to_source_path`) contains the logic that applies a code mapping to a given stack trace and returns the source path to the GH file. 

This needed to be augmented to normalize for backslashes in the stack root, as described in the PR description

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 3]
```diff
@@ -97,6 +97,80 @@ def test_raises_generic_errors(self, mock_logger):
...
 
+class TestBackSlashDeriveCodeMappings(BaseDeriveCodeMappings):
+    def setUp(self):
+        super().setUp()
+        self.platform = "python"
+        self.event_data = self.generate_data(
+            [
+                {"in_app": True, "filename": "\\sentry\\mouse.py"},
+                {"in_app": True, "filename": "\\sentry\\dog\\cat\\parrot.py"},
+                {"in_app": True, "filename": "C:sentry\\tasks.py"},
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:108]:
```suggestion
                {"in_app": True, "filename": "C:\\sentry\\tasks.py"},
```
I think this is more accurate to reality.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 4]
```diff
@@ -97,6 +97,80 @@ def test_raises_generic_errors(self, mock_logger):
...
+        repo_name = "foo/bar"
+        with patch(
+            "sentry.integrations.github.client.GitHubClientMixin.get_trees_for_org"
+        ) as mock_get_trees_for_org:
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(Repo(repo_name, "master"), ["sentry/tasks.py"])
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            assert code_mapping.stack_root == "C:sentry\\"
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:139]:
```suggestion
            assert code_mapping.stack_root == "C:\\sentry\\"
```

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 5]
```diff
@@ -97,6 +97,80 @@ def test_raises_generic_errors(self, mock_logger):
...
+        repo_name = "foo/bar"
+        with patch(
+            "sentry.integrations.github.client.GitHubClientMixin.get_trees_for_org"
+        ) as mock_get_trees_for_org:
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(Repo(repo_name, "master"), ["src/sentry/tasks.py"])
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            assert code_mapping.stack_root == "C:sentry\\"
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:154]:
```suggestion
            assert code_mapping.stack_root == "C:\\sentry\\"
```

**Review by armenzg — APPROVED** [review]
Fantastic job! 🎉 
Just a minor change for one of the cases.

File: tests/sentry/tasks/test_derive_code_mappings.py (hunk 3 above)
Comment (MichaelSun48) [reply tests/sentry/tasks/test_derive_code_mappings.py:108]: This distinction (to my surprise as well) actually exists within Windows, see [here](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats) in the official Windows docs. TL;DR is that no backslash after the drive letter signals that it is a relative path

Comment (armenzg) [reply tests/sentry/tasks/test_derive_code_mappings.py:108]: Add a note and keep it, however, it is unlikely to happen on a stack trace.

**Pull request merged**